### DISCUSSION TITLE
Websockets multiple sends won't work until recieve is called.

### DISCUSSION THREAD

**Discussion by Mithurs** [fastapi/fastapi#6988] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, WebSocket
import time

app = FastAPI()

@app.websocket("/ws")
async def doWork(ws: WebSocket):
    await ws.accept()
    ws.send_text("step 1")
    time.sleep(5)
    await ws.send_text("step 2")
    time.sleep(5)
    await ws.send_text("done")
```

### Description

* In browser terminal connect to the websocket
* Actual: After 10 seconds browser receives 3 messages ("step 1", "step 2", "done")
* Expected: browser receives "step 1" then after 5 seconds send "step 2", then after another 5 seconds send "done"
* Work around: put a `await ws.receive()` call after each send and have client send an "ack" after each message (to trigger the receive)

### Environment

* OS: Linux
* FastAPI Version: 0.63.0
* Python version: 3.8.5

**Comment by Mause** [comment, marked as answer]
You shouldn't be using time.sleep in an async function, you should use asyncio.sleep. time.sleep will block the event loop and stop network events being processed

You're already not consistently using the await operator for async calls as well, as you should be on the send_text call.

**Comment by Mithurs** [comment]
Thank you so much for the reply! That did indeed fix this problem.  However, I was trying to simplify the actual code I have which is the following:

```python
from fastapi import FastAPI, WebSocket
import time
import subprocess

app = FastAPI()

@app.websocket("/ws")
async def doWork(ws: WebSocket):
    await ws.accept()
    await callExternal()

async def callExternal(ws):
    with subprocess.Popen("<external call>", shell=True, stdout=subprocess.PIPE, bufsize=1, universal_newlines=True) as p:
        for line in p.stdout:
            await ws.send_text(line)
```
Basically the socket kicks off an external program and passes the stdout out that program line by line to the client.  In this set up it waits for the program to finish and sends all the output at once to the client.  I assume the solution will be similar to above, but I am new to python.

**Comment by Mithurs** [comment]
Upon further research the asyncio has subprocess methods to use as well and I was able to get the program running.
Thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! 👏  🙇 

Thanks for reporting back and closing the issue @Mithurs  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
